Remotion/Web/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProvider.cs

```diff
@@ -0,0 +1,251 @@
+        IResourceFileDetailsAppender resourceFileDetailsAppender,
+        IReadOnlyList<ResourceType> resourceTypes)
+    {
+      ArgumentUtility.CheckNotNull(nameof(resourcesPhysicalPathLocator), resourcesPhysicalPathLocator);
+      ArgumentUtility.CheckNotNull(nameof(resourceFileDetailsAppender), resourceFileDetailsAppender);
+      ArgumentUtility.CheckNotNull(nameof(resourceTypes), resourceTypes);
+
+      _resourcesPhysicalPathLocator = resourcesPhysicalPathLocator;
+      _resourceFileDetailsAppender = resourceFileDetailsAppender;
+      _resourceTypes = resourceTypes;
+
+      _resourceTypeLookup = resourceTypes.Select(e => e.Name).ToHashSet(StringComparer.OrdinalIgnoreCase);
+    }
+
+    public bool HasCacheKey => _cacheKey != null;
+
+    /// <inheritdoc />
+    public string GetStaticResourceCacheKey ()
+    {
+      var cacheKey = _cacheKey;
+      if (cacheKey != null)
+        return cacheKey;
+
+      lock (_lock)
+      {
+        // We ensure here that the field is only read once since it can be reset without a lock,
+        // which could cause a null value return here.
+        cacheKey = _cacheKey;
+        if (cacheKey != null)
+          return cacheKey;
+
+        if (_physicalPath == null)
+        {
+          var physicalPath = _resourcesPhysicalPathLocator.GetResourcesPhysicalPath();
+          physicalPath = Path.GetFullPath(physicalPath);
+          if (!physicalPath.EndsWith("" + Path.DirectorySeparatorChar))
+            physicalPath += Path.DirectorySeparatorChar;
+
+          _fileSystemWatcher?.Dispose();
+          _fileSystemWatcher = new FileSystemWatcher();
+          _fileSystemWatcher.Path = physicalPath;
+          _fileSystemWatcher.IncludeSubdirectories = true;
+          _fileSystemWatcher.Changed += (sender, args) => CheckForPathUpdate(args.FullPath, FileChangeType.Changed);
+          _fileSystemWatcher.Created += (sender, args) => CheckForPathUpdate(args.FullPath, FileChangeType.Created);
+          _fileSystemWatcher.Deleted += (sender, args) => CheckForPathUpdate(args.FullPath, FileChangeType.Deleted);
+          _fileSystemWatcher.Renamed += (sender, args) => CheckForPathUpdate(args.OldFullPath, FileChangeType.Renamed);
+
+          _physicalPath = physicalPath;
+          _fileSystemWatcher.EnableRaisingEvents = true;
+        }
+
+        var stringBuilder = new StringBuilder();
+        _resourceFileDetailsAppender.AppendResourceFolderDetails(stringBuilder, _physicalPath, _resourceTypes);
+
+        using var sha256 = SHA256.Create();
+        var hashBytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(stringBuilder.ToString()));
+        var hash = string.Concat(hashBytes.Select(e => e.ToString("X2")));
+
+        _cacheKey = hash;
+        return hash;
+      }
+    }
+
+    /// <inheritdoc />
+    public void Dispose ()
+    {
+      _fileSystemWatcher?.Dispose();
+      _fileSystemWatcher = null;
+    }
+
+    private void CheckForPathUpdate (string? path, FileChangeType changeType)
```

Comment:
Note: I still need to review this method. it looks okay, but I need to understand the semantics :)